crates/sui-graphql-rpc/src/metrics.rs

```diff
@@ -1,15 +1,209 @@
+        let request_metrics = RequestMetrics::new(&registry);
+
+        Self {
+            db_metrics: Arc::new(db_metrics),
+            request_metrics: Arc::new(request_metrics),
+        }
+    }
+
+    /// Updates the DB related metrics (latency, error, success)
+    pub(crate) fn observe_db_data(&self, db_latency: f64, succeeded: bool) {
+        self.db_metrics
+            .db_fetch_latency
+            .with_label_values(&["latency", "db"])
+            .observe(db_latency);
+        if succeeded {
+            self.db_metrics
+                .db_num_fetches
+                .with_label_values(&["success"])
+                .inc();
+        } else {
+            self.db_metrics
+                .db_num_fetches
+                .with_label_values(&["error"])
+                .inc();
+        }
+    }
+
+    /// Use this function to increment the number of errors due to bad input
+    pub(crate) fn inc_num_bad_input_errors(&self, label: &str) {
+        self.request_metrics
+            .num_bad_input_errors_by_path
+            // TODO when measuring cost related errors, there's no path
+            // not 100% how to label such an error?
+            .with_label_values(&[label])
+            .inc();
+    }
+
+    /// Use this function to increment the number of errors per path, and labels
+    /// them as either internal server error, bad user input, or any other error.
+    /// The path is the same as the path that the user sees.
+    pub(crate) fn inc_errors(&self, errors: Vec<ServerError>) {
+        for err in errors {
+            if let Some(ext) = err.extensions {
+                if let Some(code) = ext.get("code") {
+                    if let async_graphql_value::Value::String(val) = code.clone().into_value() {
+                        if val == code::INTERNAL_SERVER_ERROR {
+                            self.request_metrics
+                                .num_internal_errors_by_path
+                                .with_label_values(&[&self.get_query_path(err.path)])
+                                .inc();
+                        } else if val == code::BAD_USER_INPUT {
+                            self.request_metrics
+                                .num_bad_input_errors_by_path
+                                .with_label_values(&[&self.get_query_path(err.path)])
+                                .inc();
+                        } else {
+                            self.request_metrics
+                                .num_query_errors_by_path
+                                .with_label_values(&[&self.get_query_path(err.path)])
+                                .inc();
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /// Use this function to increment the number of request error.
+    /// If you need to increase error rate due to bad input, use both
+    /// this function and `inc_error_bad_input`
+    pub(crate) fn inc_num_query_success(&self) {
```

Comment:
The name of the function is query_success though. I also feel like an aggregate request error count isn't that useful. I can just sum the count of bad input and internal errors over a window on grafana

Reply:
> I can just sum the count of bad input and internal errors over a window on grafana
I'll redo this, thanks!